Framework/CurveFitting/test/Algorithms/DoublePulseFitTest.h

```diff
@@ -0,0 +1,102 @@
+            "name=ExpDecay, Height=5, Lifetime=2", 2);
+    auto doublePulseFunction =
+        getDoublePulseFunction(initialFunction, 5.0, 1.0, 2.0);
+    TS_ASSERT_EQUALS(
+        "composite=MultiDomainFunction,NumDeriv=true;(composite=Convolution,"
+        "FixResolution=false,NumDeriv=true,$domains=i;name=ExpDecay,Height=5,"
+        "Lifetime=2;(name=DeltaFunction,Height=1,Centre=-5,ties=(Height=1,"
+        "Centre=-5);name=DeltaFunction,Height=2,Centre=0,ties=(Height=2,Centre="
+        "0)));(composite=Convolution,FixResolution=false,NumDeriv=true,$"
+        "domains=i;name=ExpDecay,Height=5,Lifetime=2;(name=DeltaFunction,"
+        "Height=1,Centre=-5,ties=(Height=1,Centre=-5);name=DeltaFunction,"
+        "Height=2,Centre=0,ties=(Height=2,Centre=0)))",
+        doublePulseFunction->asString());
+  }
+
+  void
+  test_double_pulse_function_throws_exception_if_inner_function_is_composite() {
+    auto initialFunction =
+        FunctionFactory::Instance().createInitializedMultiDomainFunction(
+            "(name=ExpDecay, Height=5, Lifetime=2; name=ExpDecay, Height=7, "
+            "Lifetime=3)",
+            5);
+    TS_ASSERT_THROWS(getDoublePulseFunction(initialFunction, 5.0, 1.0, 2.0),
+                     const std::runtime_error &);
+  }
+
+  void
+  test_converting_from_double_pulse_functions_correctly_for_single_function() {
+    auto initialFunction = FunctionFactory::Instance().createInitialized(
+        "name=ExpDecay, Height=5, Lifetime=2");
+    auto doublePulseFunction = getDoublePulseFunction(
+        std::dynamic_pointer_cast<ParamFunction>(initialFunction), 5.0, 1.0,
+        2.0);
+    auto restored_function = extractInnerFunction(
+        std::dynamic_pointer_cast<Mantid::CurveFitting::Functions::Convolution>(
+            doublePulseFunction));
+
+    TS_ASSERT_EQUALS(initialFunction->asString(),
+                     restored_function->asString());
+  }
+
+  void
+  test_converting_from_double_pulse_functions_correctly_for_multidomain_function() {
+    auto initialFunction =
+        FunctionFactory::Instance().createInitializedMultiDomainFunction(
+            "name=ExpDecay, Height=5, Lifetime=2", 2);
+    auto doublePulseFunction =
+        getDoublePulseFunction(initialFunction, 5.0, 1.0, 2.0);
+    auto restored_function = extractInnerFunction(
+        std::dynamic_pointer_cast<MultiDomainFunction>(doublePulseFunction));
+
+    TS_ASSERT_EQUALS(initialFunction->asString(),
+                     restored_function->asString());
+  }
+
+  void
+  test_extracting_function_throws_exception_when_passed_non_convolution_multidomainFunction() {
+    auto initialFunction =
+        FunctionFactory::Instance().createInitializedMultiDomainFunction(
+            "(name=ExpDecay, Height=5, Lifetime=2; name=ExpDecay, Height=7, "
+            "Lifetime=3)",
+            5);
+    TS_ASSERT_THROWS(extractInnerFunction(initialFunction),
+                     const std::runtime_error &);
+  }
```

Comment:
should we test that the outputs are correct? Especially since the covariance matrix etc. are not the ones produced directly by fit.

Reply:
I have added some tests of the algorithm output on the python side.